### PR TITLE
feat: add version field to ExecutionContext

### LINKED ISSUE #1517: Add version field to ExecutionContext
Is your feature request related to a problem or challenge? Please describe what you are trying to do.

#1514 points out a secondary goal for the python library to expose the datafusion and arrow-rust versions.
I think Spark has a good implementation of this by binding the spark version to SparkContext in the jvm code, then exposing it to the pyspark API. pyspark's version itself is hard-coded. 

Describe the solution you'd like

Add a version field to the ```ExecutionContext``` struct, and expose it to the datafusion-python library. The version could then be retrieved by users by calling ```ExecutionContext.version```

Describe alternatives you've considered

Leaving it out of the rust implementation and just putting it in the python version.

Additional context
I'm not sure how arrow-rust's version would be added, but a similar approach could be taken.

### PR DESCRIPTION
**Pull request by Weijun-H**
# Which issue does this PR close?

Closes #1517

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/scheduler/task.rs [hunk 1]
```diff
@@ -279,6 +280,7 @@ impl RecordBatchStream for ExecutionResultStream {
 
 /// The shared state of all [`Task`] created from the same [`PipelinePlan`]
 #[derive(Debug)]
+#[allow(dead_code)]
```
Comment (alamb) [inline datafusion/core/src/scheduler/task.rs:283]: What do you think about adding a function like

```rust
pub fn version(&self) -> &str {
...
}
```

Which would avoid the need for allow(dead_code) perhaps

**Review by alamb — COMMENTED** [review]
Thanks @Weijun-H 

I realize https://github.com/apache/arrow-datafusion/issues/1517 says "ExecutionContext" but this code is in the experimental "scheduler"

I wonder if https://github.com/apache/arrow-datafusion/issues/1517 meant to say `SessionContext` or `SessionState` or `TaskContext` (or perhaps the code has been changed since the ticket was filed)

**Comment by Weijun-H** [comment]
> Thanks @Weijun-H
> 
> I realize #1517 says "ExecutionContext" but this code is in the experimental "scheduler"
> 
> I wonder if #1517 meant to say `SessionContext` or `SessionState` or `TaskContext` (or perhaps the code has been changed since the ticket was filed)

I vote for `SessionState`, because it is the execution context for executing queries. What do you think? 🤔

File: datafusion/core/src/execution/context.rs [hunk 2]
```diff
@@ -1803,6 +1806,11 @@ impl SessionState {
     pub fn aggregate_functions(&self) -> &HashMap<String, Arc<AggregateUDF>> {
         &self.aggregate_functions
     }
+
+    /// Return version
+    pub fn version(&self) -> &str {
+        &self.version
```
Comment (alamb) [inline datafusion/core/src/execution/context.rs:1812]: I think you could just call `env!("CARGO_PKG_VERSION")` here -- is there any reason to create a string to do so?

```suggestion
        env!("CARGO_PKG_VERSION")
```

**Review by alamb — COMMENTED** [review]
Thanks @Weijun-H !

**Review by alamb — APPROVED** [review]
Thank you @Weijun-H 👍

**Pull request merged**